src/application.rs

```diff
@@ -335,33 +335,34 @@ where
         T: FromRequest<S> + 'static,
     {
         {
-            let parts = self.parts.as_mut().expect("Use after finish");
-
-            // get resource handler
-            let mut found = false;
-            for &mut (ref pattern, ref handler) in &mut parts.resources {
-                if handler.is_some() && pattern.pattern() == path {
-                    found = true;
-                    break;
-                }
-            }
-
-            if !found {
-                let mut handler = ResourceHandler::default();
-                handler.method(method).with(f);
-                let pattern = Resource::new(handler.get_name(), path);
-                parts.resources.push((pattern, Some(handler)));
-            } else {
-                for &mut (ref pattern, ref mut handler) in &mut parts.resources {
-                    if let Some(ref mut handler) = *handler {
-                        if pattern.pattern() == path {
-                            handler.method(method).with(f);
-                            break;
+            let parts: &mut ApplicationParts<S> = self.parts.as_mut().expect("Use after finish");
+
+            let out = {
+                // get resource handler
+                let mut iterator = parts.resources.iter_mut();
+
+                loop {
+                    if let Some(&mut (ref pattern, ref mut handler)) = iterator.next() {
+                        if let Some(ref mut handler) = *handler {
+                            if pattern.pattern() == path {
+                                handler.method(method).with(f);
+                                break None;
+                            }
                         }
+                    } else {
+                        let mut handler = ResourceHandler::default();
+                        handler.method(method).with(f);
+                        let pattern = Resource::new(handler.get_name(), path);
+                        break Some((pattern, Some(handler)));
                     }
                 }
+            };
+
+            if let Some(out) = out {
```

Comment:
`Some(out)` is possible only on `iterator.next() == None` (or on the end of iterator). So you may improve the loop at line 340 like this:

```
for (pattern, handler) in iterator {
    ...
    if if pattern.pattern() == path {
        ...
        return self; // <<<<<<<<<<<<< note return self instead of break None
    }
}
// if the path was not found among handlers:
let mut handler = ResourceHandler::default();
handler.method(method).with(f);
let pattern = Resource::new(handler.get_name(), path);
parts.resources.push((pattern, Some(handler)));

return self;
```

And there is no need to create variable `let out = { huuuuge block };`

Reply:
Try to make that change and then run `cargo check`

Reply:
Thank you for pointing this out!